keras_nlp/models/whisper/whisper_audio_feature_extractor.py

```diff
@@ -0,0 +1,240 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.api_export import keras_nlp_export
+
+NUM_MELS = 80
+
+
+@keras_nlp_export("keras_nlp.models.WhisperAudioFeatureExtractor")
+class WhisperAudioFeatureExtractor(keras.layers.Layer):
+    """
+    Whisper audio feature extractor layer.
+
+    This layer takes in a batch of audio tensors, and computes the log-mel
+    spectrogram features for each audio tensor.
+
+    The input audio tensor can either be of shape `(length_of_audio,)` or
+    `(batch_size, length_of_audio)`. The output is a tensor of shape
+    `(batch_size, num_frames, NUM_MELS)`, where `num_frames` is
+    `(max_audio_length * sample_rate) / stride` and `NUM_MELS` is 80.
+
+    Args:
+        sample_rate: int, defaults to 16000. The sample rate of the audio.
+        num_fft_bins: int, defaults to 400. The size of the Fourier Transform in
+            STFT.
+        stride: int, defaults to 160. The distance between neighboring
+            sliding window frames while computing STFT.
+        max_audio_length: int, defaults to 30. The length of each audio chunk in
+            seconds. The input audio tensor will be padded/trimmed to
+            `max_audio_length*sample_rate`.
+
+    Examples:
+    ```python
+
+    # Load an audio file.
+    audio_tensor = keras_nlp.utils.audio_utils.load_audio("path/to/audio.mp3")
+
+    # Compute the log-mel spectrogram.
+    whisper_audio_feature_extractor = keras_nlp.models.WhisperAudioFeatureExtractor()
+    whisper_audio_feature_extractor(audio_tensor)
+
+    # Compute the log-mel spectrogram for a batch of audio tensors.
+    audio_tensor_1 = load_audio("path/to/audio_1.mp3")
+    audio_tensor_2 = load_audio("path/to/audio_2.mp3")
+    audio_tensor = tf.ragged.stack([audio_tensor_1, audio_tensor_2], axis=0)
+    whisper_audio_feature_extractor(audio_tensor)
+    ```
+    """
+
+    def __init__(
+        self,
+        sample_rate=16000,
+        num_fft_bins=400,
+        stride=160,
+        max_audio_length=30,
+        **kwargs,
+    ):
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
```

Comment:
I don't think we are doing anything with this right? We should only bother with this block if setting dtype will change the output dtype, but it doesn't look like that is the case today.

Reply:
I think this is important because the passed `dtype` is automatically used to typecast the input to the `call()` function. So, effectively, it controls the output dtype.

```
>>> import tensorflow as tf
>>> import keras_nlp
>>> # float32 inp, float64 output
>>> inps = tf.ones((10), dtype=tf.float32)
>>> wafe = keras_nlp.models.WhisperAudioFeatureExtractor(dtype=tf.float64)
>>> wafe(inps).dtype
tf.float64
>>> # float64 input, float32
>>> inps = tf.ones((10), dtype=tf.float64)
>>> wafe = keras_nlp.models.WhisperAudioFeatureExtractor(dtype=tf.float32)
>>> wafe(inps).dtype
tf.float32
```